docs/how-it-works.md

```diff
@@ -0,0 +1,54 @@
+# How it works
+
+The objective of this document explains how the NGINX Ingress controllers works, in particular how the NGINX model is built and why we need a one.
+
+## NGINX configuration
+
+The goal of this Ingress controller is the assembly of a configuration file (nginx.conf). This is something we need to do, at least until [this PR is merged in OpenResty](0). The main implication of this requirement is the need to reload NGINX after a change in the configuration file.
+
+## NGINX model
+
+Usually, a Kubernetes Controllers utilizes the [synchronization loop pattern](1) to check if the desired state in the controller is updated or a change is required. To this purpose, we need to build a model using different objects from the cluster, in particular (in no special order) Ingresses, Services, Endpoints, Secrets, and Configmaps to generate a point in time configuration file that reflects the state of the cluster.
+To get this object from the cluster we use [Kubernetes Informers](2), in particular `FilteredSharedInformer`. This informers allows reacting to changes in using [callbacks](3) to individual changes when a new object is added, modified or removed. Unfortunately, there is no way to know if a particular change is going to affect or not the final configuration file. This is one of the uses of the model is avoid unnecessary reloads when there's no change in the state and to detect conflicts in definitions.
+
+The final representation of the model is generated from a Go template and the model that reflect the NGINX configuration.
```

Comment:
We can provide a link to the template. We can add some information about template functions as well.

Reply:
Also it would be good to add how annotations and cm changes alter the template.